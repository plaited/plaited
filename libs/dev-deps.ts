// -- std --
export {
  assert,
  assertEquals,
  assertFalse,
  assertIsError,
  assertThrows,
} from 'https://deno.land/std@0.181.0/testing/asserts.ts'
export {
  assertSpyCall,
  assertSpyCalls,
  spy,
} from 'https://deno.land/std@0.181.0/testing/mock.ts'
export { FakeTime } from 'https://deno.land/std@0.181.0/testing/time.ts'
export { assertSnapshot } from 'https://deno.land/std@0.181.0/testing/snapshot.ts'
export { walk } from 'https://deno.land/std@0.181.0/fs/mod.ts'

// -- beautify --
export { default as beautify } from 'npm:beautify@0.0.8'

// -- esbuild --
export {
  build,
  type BuildOptions,
  type OutputFile,
  stop,
} from 'https://deno.land/x/esbuild@v0.17.12/mod.js'
export { denoPlugin } from 'https://deno.land/x/esbuild_deno_loader@0.6.0/mod.ts'
