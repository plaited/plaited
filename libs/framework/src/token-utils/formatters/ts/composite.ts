import { camelCase } from '@plaited/utils'
import { Formatter, CompositeToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
