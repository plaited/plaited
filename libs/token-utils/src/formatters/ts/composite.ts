import { camelCase } from '@plaited/utils'
import { Formatter, AliasValue, FlexToken, GridToken, TypographyToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar } from '../ts-utils.js'

type CompositeToken = FlexToken | GridToken | TypographyToken

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value as AliasValue, allTokens)}`
  }
  const toRet = Object.entries($value).map(([key, val]) => `  ${key}: ${resolveTSVar(val, allTokens)},`)
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
