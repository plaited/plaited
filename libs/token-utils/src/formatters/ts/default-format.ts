import { camelCase } from '@plaited/utils'
import { Formatter } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { isValidAlias, getCssVar } from '../ts-utils.js'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
