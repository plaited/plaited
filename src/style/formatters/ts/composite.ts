import type { CompositeToken } from '../../token.types.js'
import type { Formatter } from '../formatters.types.js'
import { camelCase } from '../../../utils/case.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
