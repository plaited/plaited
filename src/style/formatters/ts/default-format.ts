import type { Formatter } from '../formatters.types.ts'
import { camelCase } from '../../../utils/case.ts'
import { hasAlias } from '../has-alias.ts'
import { isValidAlias, getCssVar } from '../ts-utils.ts'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
