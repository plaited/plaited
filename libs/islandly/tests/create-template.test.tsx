import { assertSnapshot, assertThrows } from '../../dev-deps.ts'
import { ssr } from '../mod.ts'

Deno.test('createTemplate: self closing - html', (t) => {
  assertSnapshot(
    t,
    ssr(<input type='text'></input>),
  )
})
Deno.test('createTemplate: self closing - svg', (t) =>
  assertSnapshot(t, ssr(<polygon points='0,100 50,25 50,75 100,0'></polygon>)))

Deno.test('createTemplate: falsey - undefined', (t) =>
  assertSnapshot(t, ssr(<div>{undefined}</div>)))
Deno.test('createTemplate: falsey - null', (t) =>
  assertSnapshot(t, ssr(<div>{null}</div>)))
Deno.test('createTemplate: falsey - false', (t) =>
  assertSnapshot(t, ssr(<div>{false}</div>)))

Deno.test('createTemplate: falsey - ""', (t) =>
  assertSnapshot(t, ssr(<div>{''}</div>)))

Deno.test('createTemplate: falsey - 0', (t) =>
  assertSnapshot(t, ssr(<div>{0}</div>)))

Deno.test('createTemplate: falsey - NaN', (t) =>
  assertSnapshot(t, ssr(<div>{NaN}</div>)))

Deno.test('createTemplate: conditional', (t) =>
  assertSnapshot(t, ssr(<div>{true && 'hello'}</div>)))

Deno.test('createTemplate: array of PlaitedElements', (t) =>
  assertSnapshot(
    t,
    ssr(
      <div>
        {Array.from(Array(10).keys()).map((n) => <li>{`${n}`}</li>)}
      </div>,
    ),
  ))

Deno.test('createTemplate: should throw with attribute starting with on', () => {
  assertThrows((): void => {
    ssr(
      <div>
        <template shadowrootmode='closed'>
          <img src='nonexistent.png' onerror="alert('xss!')" />
        </template>
      </div>,
    )
  })
})

Deno.test('createTemplate: should throw on script tag', () => {
  assertThrows((): void => {
    ssr(<script type='module' src='main.js'></script>)
  })
})

Deno.test('createTemplate: should not throw on script tag with trusted attribute', (t) => {
  assertSnapshot(t, ssr(<script type='module' src='main.js' trusted></script>))
})

Deno.test('createTemplate: escapes children', (t) => {
  assertSnapshot(
    t,
    ssr(
      <div>
        {`<script type="text/javascript">
const hostRegex = /^https?:\/\/([^\/]+)\/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`}
      </div>,
    ),
  )
})

Deno.test('createTemplate: doest not escape children when trusted', (t) => {
  assertSnapshot(
    t,
    ssr(
      <div trusted>
        {`<script type="text/javascript">
const hostRegex = /^https?:\/\/([^\/]+)\/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`}
      </div>,
    ),
  )
})

Deno.test('createTemplate: Fragment of PlaitedElements', (t) =>
  assertSnapshot(
    t,
    ssr(
      <>
        {Array.from(Array(10).keys()).map((n) => <li>{`${n}`}</li>)}
      </>,
    ),
  ))

Deno.test('createTemplate: Fragment of slotted PlaitedElements', (t) =>
  assertSnapshot(
    t,
    ssr(
      <c-el>
        <slot name='slot'></slot>
        <>
          {Array.from(Array(10).keys()).map((n) => (
            <li slot='slot'>{`${n}`}</li>
          ))}
        </>
      </c-el>,
    ),
  ))
