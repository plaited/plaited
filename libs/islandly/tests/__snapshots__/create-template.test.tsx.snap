export const snapshot = {};

snapshot[`createTemplate: self closing - html 1`] = `'<input type="text"/>'`;

snapshot[`createTemplate: self closing - svg 1`] = `'<polygon points="0,100 50,25 50,75 100,0"/>'`;

snapshot[`createTemplate: falsey - undefined 1`] = `"<div></div>"`;

snapshot[`createTemplate: falsey - null 1`] = `"<div></div>"`;

snapshot[`createTemplate: falsey - false 1`] = `"<div></div>"`;

snapshot[`createTemplate: falsey - "" 1`] = `"<div></div>"`;

snapshot[`createTemplate: falsey - 0 1`] = `"<div></div>"`;

snapshot[`createTemplate: falsey - NaN 1`] = `"<div></div>"`;

snapshot[`createTemplate: conditional 1`] = `"<div>hello</div>"`;

snapshot[`createTemplate: style attribute 1`] = `'<div style="background-color:blue; margin:12px;">styles</div>'`;

snapshot[`createTemplate: data-trigger attribute 1`] = `'<div data-trigger="click->random focus->thing">triggers</div>'`;

snapshot[`createTemplate: array of PlaitedElements 1`] = `"<div><li>0</li><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li></div>"`;

snapshot[`createTemplate: should not throw on script tag with trusted attribute 1`] = `'<script type="module" src="main.js"></script>'`;

snapshot[`createTemplate: escapes children 1`] = `
"<div>&lt;script type=&quot;text/javascript&quot;&gt;
const hostRegex = /^https?://([^/]+)/.*\$/i;
const host = document.URL.replace(hostRegex, &#39;\$1&#39;);
const socket = new WebSocket(/);
const reload = () =&gt;{
  location.reload();
  console.log(&#39;...reloading&#39;);
};
socket.addEventListener(&#39;message&#39;, reload);
console.log(&#39;[plaited] listening for file changes&#39;);
&lt;/script&gt;</div>"
`;

snapshot[`createTemplate: doest not escape children when trusted 1`] = `
\`<div><script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*\$/i;
const host = document.URL.replace(hostRegex, '\$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script></div>\`
`;

snapshot[`createTemplate: Fragment of PlaitedElements 1`] = `"<li>0</li><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li>"`;

snapshot[`createTemplate: with slotted PlaitedElements 1`] = `'<c-el><template shadowrootmode="open" shadowrootdelegatesfocus="true"><slot name="slot"></slot></template><li slot="slot">slot-0</li><li slot="slot">slot-1</li><li slot="slot">slot-2</li><li slot="slot">slot-3</li><li slot="slot">slot-4</li><li slot="slot">slot-5</li><li slot="slot">slot-6</li><li slot="slot">slot-7</li><li slot="slot">slot-8</li><li slot="slot">slot-9</li></c-el>'`;

snapshot[`createTemplate: Fragment PlaitedElements 1`] = `"<li>item-0</li><li>item-1</li><li>item-2</li><li>item-3</li><li>item-4</li><li>item-5</li><li>item-6</li><li>item-7</li><li>item-8</li><li>item-9</li>"`;
